Improvements to trackedFunction and State

lolmaus opened the pull request:
Here are the changes:

* Renamed the `isResolved` state flag to `isFinished` to avoid ambiguity. `isResolved` seems related to `resolvedValue`, so one would expect `resolvedValue` to be populated when `isResolved` is true.
* Added an `isSuccessful` state flag.
* Wrapped the whole `retry` into `waitForPromise()`.
* `retry` now resets state flags eagerly. Without this change, it was impossible to tell if a retried `trackedFunction` resource is pending or not.
* Added `previousResovledValue` property and `previousValue` getter, so that the original data can remain on the screen while a resource is being retried or has errored after an unsuccessful retry.

    I chose not to reuse the `resolvedValue` property to maintain the previous value. This way, keeping the original data on the screen requires some more boilerplate code (e. g. `{{or this.resource.value this.resource.previousValue}}`), but the benefit is that the behaviour is not surprising to developers unfamiliar with the previous value feature.
* When arguments of a  `trackedFunction` resource change, it would reinstantiate `State`. Previously, it resulted in the loss of previous state. Now, the previous state is persisted across argument changes.
* Since `State` does not have a `hasNeverStarted` flag, it now performs a `this.retry()` in the constructor. It is now impossible to have a non-pending, non-complete `State`.

@NullVoxPopuli If you approve the general direction of this PR, I would like to add more tests: cover the whole functionality of `State` in a separate test file. I would also like to add some TypeDoc documentation.

PS The `async functions update when the promise resolves` test is flaky: it consistently passes for me locally if the browser window is focused and consistently fails otherwise.

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
  * State container that represents the asynchrony of a `trackedFunction`
  */
 export class State<Value> {
-  @tracked isResolved = false;
```
can we have isResolved and isSuccessful together? -- we can't actually get rid of isResolved (or any other property), as that is a breaking change.

lolmaus replied:
What do you mean "have them together"?

They mean different things. Originally, whent he promise was rejected, `isResolved` would become `true` ([source](https://github.com/NullVoxPopuli/ember-resources/blob/824404911d/ember-resources/src/util/function.ts#L193-L195))).

Since it was very counter intuitive to have `isResolved` as `true` when the promise was in fact not resolved, I chose to remove this flag and replace with unambiguous ones: `isSuccessful` and `isFinished`, borrowing from Ember Concurrency.

I believe that this is reason is very worthy of a breaking change and a major version bump.

If you disagree, I can make `isResolved` a getter alias. But note that it will have to be an alias to `isFinished` and not to `isResolved` as you say in the previous comment.

NullVoxPopuli replied:
> What do you mean "have them together"?

Like this:

> I can make isResolved a getter alias. But note that it will have to be an alias to isFinished and not to isResolved as you say in the previous comment.

sounds good to me!

>  If you disagree, 

yeah -- mostly wanting to avoid a major bump atm

> I believe that this reason is very worthy of a breaking change and a major version bump.

unfortunately, `trackedFunction` is versioned with `ember-resources`, and there have been no significant enough changes to warrant a major bump. There are some things I'd like to explore, and we can plan out what we want for a major, but idk how soon we'd want to release that.
(things I've been wanting to see implemented soon: registerable usables, a resource blueprint, etc)

lolmaus replied:
>> I can make isResolved a getter alias. But note that it will have to be an alias to isFinished and not to isResolved as you say in the previous comment.

> sounds good to me!

Will do.

***

> unfortunately, trackedFunction is versioned with ember-resources, and there have been no significant enough changes to warrant a major bump.

Please excuse me for starting a holy war, but I believe that SemVer goal is to capture breaking changes, however small, and NOT to capture the significance of changes. A spiking major version number is an indication of proper SemVer usage.

I will add a alias `isResolved`, but I object: this property will become `true` when the promise is not resolved. This should be fixed, and there is no point in holding to a major version number.

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
     this.#hooks = hooks;
     this.#initialValue = initialValue;
+    this.previousResolvedValue = previousResolvedValue;
+    this.retry();
```
why move retry here from the the resource body, where it used to be?

I know it's functionally the same (afaict, anyway), but from a side-effect and readability perspective, I think it makes sense to not start promised-based things in constructors

lolmaus replied:
My reasoning was that `State` could be imported directly and when instantiated without `retry` it would be in an unsupported state.

I can see three solutions to this:

1. Undo this change and embrace the unsupported state.
2. Make the initial state supported by adding an `isInitial` or `hasNotStarted` flag.
3. Wrap `this.retry()` with an `if` contional controlled by a constructor argument. This would allow instantiating it without retrying only consciously, e. g. in tests.

NullVoxPopuli replied:
I'm kinda leaning towards a secret 4th option, where maybe we make `State` an interface, and not allow instantiation of the `State` class at all? thoughts? (It was probably a bug that `State` is importable and usable as more than a type anyway)

lolmaus replied:
Hmm, I like the idea of strict immutable objects, but I don't see how I can implement this. The logic of creating those immutable objects must go somewhere, and it should not be `trackedFunction` itself.

The beauty of the `State` class is that it is a single-responsibility entity that deals with asynchrony. I can see `State` spit out `ImmutableState`s, but I don't see how I can replace `State` with `ImmutableState`. 

Please explain and I'm happy to work on it.

lolmaus replied:
(Out of the first three options, I think the second one is the most reasonable.)

NullVoxPopuli replied:
Resolving, as I think we've kinda talking about the same thing in two places -- here: https://github.com/NullVoxPopuli/ember-resources/pull/683#discussion_r1038506316

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
     return this.isPending;
   }
 
-  get isError() {
```
we still need isError

lolmaus replied:
Responded in another comment thread.

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
+
+  protected _retry = async () => {
+    try {
+      // We need to invoke this before going async so that tracked properties are consumed (entangled with) synchronously
```
thanks for the comments!! <3 :tada:

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
+        this.previousResolvedValue = this.resolvedValue;
+      }
+
+      this.isSuccessful = false;
```
shouldn't isSuccessful be a getter based on not having an error?

lolmaus replied:
I initially made the `resolvedValue` and `error` properties the source of truth: when one of them stops being undefined, then the promise has finished.

The problem is that the promise can resolve with `undefined` which will result in an inconsistent state of `State`. Yes, the JSON spec does not have `undefined`, but network requests aren't the only use case for `trackedFunction` and `State`. The developer can choose to `resolve()` or `reject()` their promise without arguments, which resolves/rejects with `undefined`. That's legit JavaScript, so `State` should support it.

Thus, I had to make state flags simple properties. Since I went with boolean properties, and there are three unique states (pending/successful/failed), I had to employ two boolean properties.

An alternative would be to have a single simple property to store promise state, e. g.:

```ts
type PromiseState = 'pending' | 'successful' | 'error';
```

or

```ts
enum PromiseState { Pending, Successful, Error }
```

Then `isSusccessful` and `isError` can be getters.

***

Please tell me which option you prefer.

NullVoxPopuli replied:
Yeah, this makes sense -- and now I'm kinda wondering if we need to approach `State` differently, like define a stricter set of states that `State` can be in, like:

```ts
type InternalState = 
 | {
     status: 'pending';
   }
 | {
     status: 'error';
     error: Error;
   }
 | {
   status: 'successful';
   value: /* whatever this would be */
   }  
```

Then, we can be reasonably certain there is no possible overlap between the InternalStates?
thoughts?

NullVoxPopuli replied:
your idea of implementing the `State` class is 💯 I only mean that the return type be different -- rather than exposing the class as the return type, we annotate the return type to be this `InternalState`

NullVoxPopuli commented on ember-resources/src/util/function.ts:
```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
+      }
+
+      this.isSuccessful = false;
+      this.isError = false;
```
why wouldn't we want isError to be a getter?

NullVoxPopuli reviewed (changes requested):
Adding previous value state looks great!

Left some comments / questions / constraints. 

> @NullVoxPopuli If you approve the general direction of this PR, I would like to add more tests: cover the whole functionality of State in a separate test file. I would also like to add some TypeDoc documentation.

Yeah, I think the overall direction / new features looks good -- I think what we both want is achievable! 


> When arguments of a trackedFunction resource change, it would reinstantiate State. Previously, it resulted in the loss of previous state. Now, the previous state is persisted across argument changes.

I wonder if we should have a more complex "previous" state -- like, what happens if the previous state was an error?

> Since State does not have a hasNeverStarted flag, it now performs a this.retry() in the constructor. It is now impossible to have a non-pending, non-complete State.

was this a possible state to be in? upon access of any property the `retry` method should have been called. :thinking: 
I wonder if the initial values of the properties maybe had the wrong defaults?

PS The async functions update when the promise resolves test is flaky: it consistently passes for me locally if the browser window is focused and consistently fails otherwise.

lolmaus commented:
Hi! Thanks for looking into this. I'll try to be more responsive from now on.

> I wonder if we should have a more complex "previous" state -- like, what happens if the previous state was an error?

My current implementation does not persist all `State`'s state, but rather only `this.resolvedValue || this.previousResolvedValue`. 

My reasoning is that there is no practical case to persist everything. Persisting the resolved value is necessary to avoid the original data to be removed from the screen when the users refreshes. A loading spinner could be overlaid on top of old data or something. This would prevent a scrollbar jump and let the user read the original data while he's waiting for fresh data.

Whereas keeping the error message on the screen after the user has started a retry — has no value.

Note that if the user loaded the data successfully (promise 1 successful), then attempted a refresh which failed (promise 2 failed), then attempted another refresh (promise 3 pending) — then the resource will keep showing last succesful data from promise 1. It will not replace the data from promise 1 with an error from promise 2.

***

>> Since State does not have a hasNeverStarted flag, it now performs a this.retry() in the constructor. It is now impossible to have a non-pending, non-complete State.

> was this a possible state to be in? upon access of any property the retry method should have been called. 🤔

It was not possible to enter this state when using `trackedFunction` because it was doing `retry()` right after `new`. But it was possible to import `State`, instantiate it — and have it in an unsupported state.

> I wonder if the initial values of the properties maybe had the wrong defaults?

It's not the case of wrong defaults. It's the case of not having an `isInitial` flag or something like that. `hasNotStarted`?

***

>> PS The async functions update when the promise resolves test is flaky: it consistently passes for me locally if the browser window is focused and consistently fails otherwise.

> hmmm. window focus-based tests are rough.
> any chance you know what's going on with that one?

There are two problems with this test:

* It relies on knowledge of internal implementation to execute steps. It knows when to use and when not use `settled()` in order to work around `waitForPromise()` which would otherwise prevent the test from asserting the intermediate state.
* It relies on a magic number of milliseconds to wait for, namely [25](https://github.com/NullVoxPopuli/ember-resources/blob/8244049/testing/ember-app/tests/utils/function/rendering-test.ts#L102). When a browser tab is unfocused, the browser throttles JavaScript execution in it, and test steps takes longer than 25 ms.

So increasing the number of milliseconds should resolve the issue. But it's not a graceful solution. It's balancing between slow and flaky.

A good solution would be to employ something like @rwjblue's [ember-lifeline](https://ember-lifeline.github.io/ember-lifeline/docs/why-use-ember-lifeline) to create awaitable stop points that let you manually advance through in the test. But this addon makes the code substantally more complicated.

So my preference here is simply not to test the intermediate state.

lolmaus commented:
@NullVoxPopuli, I've added test coverage for `State` and it revealed unexpected behavior of the `value` getter (and `previousValue` which I created by borrowing from `value`) in combination with `initialValue`:

if you pass `initialValue` to `State` and its promise fails, then `value` would return the initial value. This may be very counterintuitive, as developers might consider a resource to be successful when its `value` is truthy.

NullVoxPopuli commented:
>  I've added test coverage for State

most excellent, thank you!!!

> My current implementation does not persist all State's state, but rather only this.resolvedValue || this.previousResolvedValue.

do you think there is a way to combine `keepLatest` to achieve this same behavior?
I worry that by forcing certain patterns in the `trackedFunction` implementation, we eliminate some flexibility. Idk what everyone _could_ want to do, but enabling patterns through composition feels more flexible to me? idk -- thoughts?
Here is a [demo of using trackedFunction + keepLatest](https://limber.glimdown.com/edit?format=glimdown&t=%23%20%60trackedFunction%60%20%2B%20%60keepLatest%60%0A%0AUses%20the%20StarWars%20API%0A%0A%60%60%60gjs%20live%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aimport%20%7B%20use%20%7D%20from%20%27ember-resources%27%3B%0Aimport%20%7B%20trackedFunction%20%7D%20from%20%27ember-resources%2Futil%2Ffunction%27%3B%0Aimport%20%7B%20keepLatest%20%7D%20from%20%27ember-resources%2Futil%2Fkeep-latest%27%3B%0A%0Aconst%20urlFor%20%3D%20(id)%20%3D%3E%20%60https%3A%2F%2Fswapi.dev%2Fapi%2Fpeople%2F%24%7Bid%7D%60%3B%0A%0Aexport%20default%20class%20Demo%20extends%20Component%20%7B%0A%20%20%40tracked%20id%20%3D%2051%3B%0A%20%20updateId%20%3D%20(event)%20%3D%3E%20this.id%20%3D%20event.target.value%3B%0A%0A%20%20request%20%3D%20trackedFunction(this%2C%20async%20()%20%3D%3E%20%7B%0A%20%20%20%20let%20response%20%3D%20await%20fetch(urlFor(this.id))%3B%0A%20%20%20%20let%20json%20%3D%20await%20response.json()%3B%0A%20%20%20%20return%20json%3B%0A%20%20%7D)%3B%0A%0A%20%20%40use%20data%20%3D%20keepLatest(%7B%0A%20%20%20%20value%3A%20()%20%3D%3E%20this.request.value%20%3F%3F%20%7B%7D%2C%0A%20%20%20%20when%3A%20()%20%3D%3E%20this.request.isLoading%2C%0A%20%20%7D)%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%3Cdiv%20class%3D%22border%20p-4%20grid%20gap-4%22%3E%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%7B%7Bthis.data.name%7D%7D%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.request.isLoading%7D%7D%0A%20%20%20%20%20%20%20%20%20%20...%20loading%20%7B%7B%40id%7D%7D%20...%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20Person%20ID%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27number%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20class%3D%27border%20px-3%20py-2%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20value%3D%7B%7Bthis.id%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27input%27%20this.updateId%7D%7D%20%2F%3E%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%60%60%60%0A)

> But it was possible to import State, instantiate it — and have it in an unsupported state.

This should probably not be supported. The type is public, but instantiation of the `State` is not supported (or rather, I don't want to support that use case, because I don't think it makes sense).

>  It's the case of not having an isInitial flag or something like that. hasNotStarted?

given the above, I don't see how there is value is these flags? what's the use case? (Just trying to keep public API as small as possible, ya know?)

> So my preference here is simply not to test the intermediate state.

we could probably substitute with assert.step  -- but I think there are tests already do that -- maybe we do just delete that test? (I'm cool with that if you want to remove it in this PR, the extra stability of the test suite from doing so is more than welcome!)

>  This may be very counterintuitive, as developers might consider a resource to be successful when its value is truthy.

good thing we have isSuccessful and isResolved  to differentiate :upside_down_face:

NullVoxPopuli commented on testing/ember-app/tests/utils/function/state-test.ts:
```diff
@@ -0,0 +1,447 @@
+import { settled } from '@ember/test-helpers';
```
thanks for all this! :tada: <3

lolmaus commented:
> do you think there is a way to combine keepLatest to achieve this same behavior?

Oh, I wasn't aware about `@keepLatest`! I didn't even realise that resources could be composed like this. This is awesome!

Yes, `@keepLatest` fully removes the necessity in `previousResolvedValue`. 👍

In theory.

In practice though, the current implementation has two flaws from the perspective of my use case:
1. It forgets the previous value when the dependency resource rejects.
2. It obscures access to the error, so in order to implement the UI, I would need to fiddle with both the `keepLatest` resource and the dependency resource.

To clarify my use case and expectation, consider this sequence of events:

1. Start promise 1.
2. Promise 1 resolves with value `"foo"`.
3. Start promise 2.
4. Promise 2 rejects with an error.
5. Start promise 3.

I want the `"foo"` value to be available on steps 3 to 5. The current `keepLatest` implementation makes `"foo"` available only on step 3.

So my suggestion is that I remove the previous value functionality from this PR, get it merged, then start another PR to update `keepLatest`.

***

BTW, the API docs for `keepLatest` are broken: https://ember-resources.pages.dev/modules/util_keep_latest

The Limber demo you linked to is not working either, because of the CORS error when accessing a third-party API.

I have updated the demo not to use the failing API and to be able to reject, please have a look: [demo](https://limber.glimdown.com/edit?format=glimdown&t=%23%20%60trackedFunction%60%20%2B%20%60keepLatest%60%0A%0A%60%60%60gjs%20live%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aimport%20%7B%20use%20%7D%20from%20%27ember-resources%27%3B%0Aimport%20%7B%20trackedFunction%20%7D%20from%20%27ember-resources%2Futil%2Ffunction%27%3B%0Aimport%20%7B%20keepLatest%20%7D%20from%20%27ember-resources%2Futil%2Fkeep-latest%27%3B%0A%0Aconst%20urlFor%20%3D%20(id)%20%3D%3E%20%60https%3A%2F%2Fswapi.dev%2Fapi%2Fpeople%2F%24%7Bid%7D%60%3B%0A%0Aexport%20default%20class%20Demo%20extends%20Component%20%7B%0A%20%20%40tracked%20id%20%3D%2051%3B%0A%20%20shouldFail%20%3D%20false%3B%0A%20%20updateId%20%3D%20(event)%20%3D%3E%20this.id%20%3D%20event.target.value%3B%0A%20%20updateShouldFail%20%3D%20(event)%20%3D%3E%20this.shouldFail%20%3D%20event.target.checked%3B%0A%0A%20%20request%20%3D%20trackedFunction(this%2C%20async%20()%20%3D%3E%20%7B%0A%20%20%20%20this.id%3B%20%2F%2F%20Consuming%20id%2C%20so%20that%20the%20function%20restarts%20on%20id%20change%0A%20%20%20%20return%20new%20Promise((resolve%2C%20reject)%20%3D%3E%20%7B%20%20%20%20%0A%20%20%20%20%20%20setTimeout(()%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20if%20(this.shouldFail)%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20reject%27)%0A%20%20%20%20%20%20%20%20%20%20reject(new%20Error(%27Oops!%27))%3B%0A%20%20%20%20%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20resolve%27)%0A%20%20%20%20%20%20%20%20%20%20resolve(this.id)%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%2C%201000)%3B%0A%20%20%20%20%7D)%3B%0A%20%20%7D)%3B%0A%0A%20%20%40use%20data%20%3D%20keepLatest(%7B%0A%20%20%20%20value%3A%20()%20%3D%3E%20this.request.value%2C%0A%20%20%20%20when%3A%20()%20%3D%3E%20this.request.isLoading%2C%0A%20%20%7D)%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%3Cdiv%20class%3D%22border%20p-4%20grid%20gap-4%22%3E%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20Resolved%20value%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bthis.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.request.isLoading%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20...%20loading%20%7B%7B%40id%7D%7D%20...%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27checkbox%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20checked%3D%7B%7Bthis.shouldFail%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27change%27%20this.updateShouldFail%7D%7D%20%2F%3E%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20Should%20fail%20%20%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20Person%20ID%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27number%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20class%3D%27border%20px-3%20py-2%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20value%3D%7B%7Bthis.id%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27input%27%20this.updateId%7D%7D%20%2F%3E%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%60%60%60%0A).

***

> This should probably not be supported. The type is public, but instantiation of the State is not supported (or rather, I don't want to support that use case, because I don't think it makes sense).

> given the above, I don't see how there is value is these flags? what's the use case? (Just trying to keep public API as small as possible, ya know?)

I think that stateful `State` class is a very useful concept and can be used outside of `trackedFunction`.

I believe that JavaScript should have `State` functionality built into `Promise`, and since it doesn't, the world needs a shim!

Thus, it should be importable, it shoudl be usable as a standalone class, and the `hasNotStarted` flag is necessary to make it consistent.

***

> we could probably substitute with assert.step -- but I think there are tests already do that -- maybe we do just delete that test? (I'm cool with that if you want to remove it in this PR, the extra stability of the test suite from doing so is more than welcome!)

I'll remove the intemrediate assertions and keep the test with stable assertions only.

UPD: There's something fishy about that test. It fails not only on intermetdiate assertions after `await timeout(25)`, but on what I considered stable assertions — after `await settled()`.

The test stabilizes if I do `await settled(); await settled()`.  This may be a consequence of using [this trick](https://github.com/lolmaus/ember-resources/blob/71438f0630f0364ac3cd3fa8ae6a38a66dd29331/ember-resources/src/util/function.ts#L191), which I didn't want to add, but I found no other way to reset flags without triggering the "was modified twice in a single render" error.

***

>> This may be very counterintuitive, as developers might consider a resource to be successful when its value is truthy.

> good thing we have isSuccessful and isResolved to differentiate 🙃

Not good enough! 😬

NullVoxPopuli commented:
> I have updated the demo not to use the failing API and to be able to reject, please have a look: [demo](https://limber.glimdown.com/edit?format=glimdown&t=%23%20%60trackedFunction%60%20%2B%20%60keepLatest%60%0A%0A%60%60%60gjs%20live%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aimport%20%7B%20use%20%7D%20from%20%27ember-resources%27%3B%0Aimport%20%7B%20trackedFunction%20%7D%20from%20%27ember-resources%2Futil%2Ffunction%27%3B%0Aimport%20%7B%20keepLatest%20%7D%20from%20%27ember-resources%2Futil%2Fkeep-latest%27%3B%0A%0Aconst%20urlFor%20%3D%20(id)%20%3D%3E%20%60https%3A%2F%2Fswapi.dev%2Fapi%2Fpeople%2F%24%7Bid%7D%60%3B%0A%0Aexport%20default%20class%20Demo%20extends%20Component%20%7B%0A%20%20%40tracked%20id%20%3D%2051%3B%0A%20%20shouldFail%20%3D%20false%3B%0A%20%20updateId%20%3D%20(event)%20%3D%3E%20this.id%20%3D%20event.target.value%3B%0A%20%20updateShouldFail%20%3D%20(event)%20%3D%3E%20this.shouldFail%20%3D%20event.target.checked%3B%0A%0A%20%20request%20%3D%20trackedFunction(this%2C%20async%20()%20%3D%3E%20%7B%0A%20%20%20%20this.id%3B%20%2F%2F%20Consuming%20id%2C%20so%20that%20the%20function%20restarts%20on%20id%20change%0A%20%20%20%20return%20new%20Promise((resolve%2C%20reject)%20%3D%3E%20%7B%20%20%20%20%0A%20%20%20%20%20%20setTimeout(()%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20if%20(this.shouldFail)%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20reject%27)%0A%20%20%20%20%20%20%20%20%20%20reject(new%20Error(%27Oops!%27))%3B%0A%20%20%20%20%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20resolve%27)%0A%20%20%20%20%20%20%20%20%20%20resolve(this.id)%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%2C%201000)%3B%0A%20%20%20%20%7D)%3B%0A%20%20%7D)%3B%0A%0A%20%20%40use%20data%20%3D%20keepLatest(%7B%0A%20%20%20%20value%3A%20()%20%3D%3E%20this.request.value%2C%0A%20%20%20%20when%3A%20()%20%3D%3E%20this.request.isLoading%2C%0A%20%20%7D)%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%3Cdiv%20class%3D%22border%20p-4%20grid%20gap-4%22%3E%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20Resolved%20value%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bthis.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.request.isLoading%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20...%20loading%20%7B%7B%40id%7D%7D%20...%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27checkbox%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20checked%3D%7B%7Bthis.shouldFail%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27change%27%20this.updateShouldFail%7D%7D%20%2F%3E%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20Should%20fail%20%20%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20Person%20ID%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27number%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20class%3D%27border%20px-3%20py-2%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20value%3D%7B%7Bthis.id%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27input%27%20this.updateId%7D%7D%20%2F%3E%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%60%60%60%0A).

thanks!!

> I want the "foo" value to be available on steps 3 to 5. The current keepLatest implementation makes "foo" available only on step 3.

I think [this fixed it](https://limber.glimdown.com/edit?format=glimdown&t=%23%20%60trackedFunction%60%20%2B%20%60keepLatest%60%0A%0A%60%60%60gjs%20live%0Aimport%20Component%20from%20%27%40glimmer%2Fcomponent%27%3B%0Aimport%20%7B%20tracked%20%7D%20from%20%27%40glimmer%2Ftracking%27%3B%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0A%0Aimport%20%7B%20use%20%7D%20from%20%27ember-resources%27%3B%0Aimport%20%7B%20trackedFunction%20%7D%20from%20%27ember-resources%2Futil%2Ffunction%27%3B%0Aimport%20%7B%20keepLatest%20%7D%20from%20%27ember-resources%2Futil%2Fkeep-latest%27%3B%0A%0Aconst%20urlFor%20%3D%20(id)%20%3D%3E%20%60https%3A%2F%2Fswapi.dev%2Fapi%2Fpeople%2F%24%7Bid%7D%60%3B%0A%0Aexport%20default%20class%20Demo%20extends%20Component%20%7B%0A%20%20%40tracked%20id%20%3D%2051%3B%0A%20%20shouldFail%20%3D%20false%3B%0A%20%20updateId%20%3D%20(event)%20%3D%3E%20this.id%20%3D%20event.target.value%3B%0A%20%20updateShouldFail%20%3D%20(event)%20%3D%3E%20this.shouldFail%20%3D%20event.target.checked%3B%0A%0A%20%20request%20%3D%20trackedFunction(this%2C%20async%20()%20%3D%3E%20%7B%0A%20%20%20%20this.id%3B%20%2F%2F%20Consuming%20id%2C%20so%20that%20the%20function%20restarts%20on%20id%20change%0A%20%20%20%20return%20new%20Promise((resolve%2C%20reject)%20%3D%3E%20%7B%20%20%20%20%0A%20%20%20%20%20%20setTimeout(()%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20if%20(this.shouldFail)%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20reject%27)%0A%20%20%20%20%20%20%20%20%20%20reject(new%20Error(%27Oops!%27))%3B%0A%20%20%20%20%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%20%20%20%20%20%20console.log(%27getData%20resolve%27)%0A%20%20%20%20%20%20%20%20%20%20resolve(this.id)%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%2C%201000)%3B%0A%20%20%20%20%7D)%3B%0A%20%20%7D)%3B%0A%0A%20%20%40use%20data%20%3D%20keepLatest(%7B%0A%20%20%20%20value%3A%20()%20%3D%3E%20this.request.value%2C%0A%20%20%20%20when%3A%20()%20%3D%3E%20this.request.isLoading%20%7C%7C%20this.request.isError%2C%0A%20%20%7D)%3B%0A%0A%20%20%3Ctemplate%3E%0A%20%20%20%20%3Cdiv%20class%3D%22border%20p-4%20grid%20gap-4%22%3E%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20Resolved%20value%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bthis.data%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%7B%7B%23if%20this.request.isLoading%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%3Cp%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20...%20loading%20%7B%7B%40id%7D%7D%20...%0A%20%20%20%20%20%20%20%20%20%20%3C%2Fp%3E%0A%20%20%20%20%20%20%20%20%7B%7B%2Fif%7D%7D%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27checkbox%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20checked%3D%7B%7Bthis.shouldFail%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27change%27%20this.updateShouldFail%7D%7D%20%2F%3E%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20Should%20fail%20%20%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%0A%20%20%20%20%20%20%20%20%3Clabel%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20Person%20ID%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cinput%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20type%3D%27number%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20class%3D%27border%20px-3%20py-2%27%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20value%3D%7B%7Bthis.id%7D%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7B%7Bon%20%27input%27%20this.updateId%7D%7D%20%2F%3E%0A%20%20%20%20%20%20%20%20%3C%2Flabel%3E%0A%20%20%20%20%3C%2Fdiv%3E%0A%20%20%3C%2Ftemplate%3E%0A%7D%0A%60%60%60%0A)?


> BTW, the API docs for keepLatest are broken: https://ember-resources.pages.dev/modules/util_keep_latest

yeah, I think typedoc broke, or I broke, I haven't looked in to it yet :(  

> The Limber demo you linked to is not working either, because of the CORS error when accessing a third-party API.

it _did_ used to work. the CORS changes must be new -- and make me sad, because there are so few APIs you can actually use in a REPL :( 


> I think that stateful State class is a very useful concept and can be used outside of trackedFunction.
> I believe that JavaScript should have State functionality built into Promise, and since it doesn't, the world needs a shim!
> Thus, it should be importable, it should be usable as a standalone class, and the hasNotStarted flag is necessary to make it consistent.

ah, I see why you've been motivated for some of these changes now -- this makes sense. And I largely agree (built in Promise is very lacking) -- however, I don't think _I_ want to be responsible for such an abstraction (I could be convinced -- like, if you want to be a regular maintainer, then that'd help quite a bit <3 ).

I will gladly consume some other library's public API that provides its own `Stateful` thing -- maybe even this one: https://github.com/chriskrycho/ember-async-data/blob/main/ember-async-data/src/tracked-async-data.ts

lolmaus commented:
Hey @NullVoxPopuli, I'm sorry for going silent. I'm back.

I've looked into `TrackedAsyncData` from `ember-async-data` and it's absolutely awesome. So much so that it effectively eliminates the need in `trackedFunction` and `State`! 🙀 

`TrackedAsyncData` has all the flags and also takes care of the destroyable.

The only things it lacks compared to `ember-resources` are:

1. Retrying. But it can be achieved like this:

    ```js
    @tracked promise = this.myService.getData(); // Sync

    @cached
    get data() {
      return new TrackedAsyncData(this.promise, this);
    }

    restart = (shouldFail = false) => {
      this.promise = this.myService.getData(); // Sync
    };
    ```

    In this example, I instantiate the promise outside of the cached getter that instantiates `TrackedAsyncData`. This makes `TrackedAsyncData` reactive.

2. Conciseness. The code has much more boilerplate compared to `trackedFunction`:

    ```js
    resource = trackedFunction(this, () => {
      return this.myService.getData();
    });
    ```

    Instantiating the promise by hand feels less declarative and more verbose than what `ember-resources` does, but also less magical and more obvious.

3. Keeping the previous value when retrying. But as you said, it's better handled by `keepLatest`.

***

I see the following alternative directions to proceed with:

1. Close this PR and keep `trackedFunction` and `State` as is.

    I don't like this because `trackedFunction` and `State` have multiple issues mentioned in the PR description. They feel kinda unfinished.

2. Finalize this PR in its current direction.

    I will need your advise with how to proceed, while avoiding overlap with `TrackedAsyncData`. We need to decide where to draw the line.

3. Throw out `State` and replace it with `TrackedAsyncData`.

    This would be a consice, declarative, entangling wrapper for `ember-async-data`.

    Please share your thoughts on how to integrate them together, without making changes to `TrackedAsyncData`. We need to find a new way to offer retrying. We could subclass `TrackedAsyncData`, but maybe we could do it without tight coupling between `TrackedAsyncData` and `trackedFunction`?

    This would be a breaking change, since flags have different names, but that's a good thing because `TrackedAsyncData` has no ambiguity of making `isResolved` true for a rejected promise.

lolmaus commented:
@NullVoxPopuli I have replaced the promise access functionality with `TrackedData` from `ember-async-data`.

1. I failed to replace `State` with `TrackedData` because `TrackedData` is implemented with weird TypeScript hacks that make it impossible to subclass. :( Intentionally! ☹ So I had to put `TrackedData` inside `State` rather than make `State` a subcalss of `TrackedData`.

2. I didn't want to access promise values as `this.myResource.asyncData.isPending`, so I provided getters on `State` that delegate to `TrackedData`, and the follwoing is possible: `this.myResource.isPending`.

3. I chose to stay true to `TrackedData`'s footprint as it is unambiguous, well-typed and kinda standard. There are no aliases such as `get isFinished { return this.isResovled; }` because KISS. There is no ambiguity with `isResolved` being truthy for a rejected promise.

    Yes, this is a breaking change, but it's a necessary breaking change. WIth SemVer, you're not supposed to cling to a major version number and avoid bumping it for a small breaking change. You should bump it as often as it needs. SemVer is for package managers, not for people.

5. Unfortunately, I was unable to use promise value types from `ember-async-data` (the `Stateful` interface that you had suggested) because those types extend the whole `TrackedData` and are not a simple interface. 👎

6. I removed `retry()` from the `State`'s constructor, so it needs to be called explicitly when `State` is used separately.  This is for two reasons:
    * JavaScript does not allow for constructors to be async, and starting a promise from a constructor is just wrong. You can't await it, you can't get a reference.
    * As a result, the rejection test was crashing due to an uncaught promise. Unfortunately, I could not catch the existing promise in the test because the `async` keyword generates a new promise even when you `return` a promise. With explicit `retry`, I was able to do `const promise2 = state.retry()` and `try { await promise2} catch(e)`.

7. I took liberty to remove the `initialState` feature from `trackedFunction` and `State`. As I proceeded to unit-test `State`, I had to specify all possible combinations of all getters. I discovered that `myResource.value` behaved unexpectedly.

    For example, a promise rejects, but `myResource.value` is truthy. I believe it is very common for developers to do things like:

        {{#if this.resource.value}}
          Succcess: {{this.resource.value}}
        {{else if this.resource.isPending}}
          ...

    The initial value feature breaks this pattern in an unintuitive way.

    I gave it a lot of thought trying to rationalize the behavior of `myResource.value` in different states, given `initialState`. And I failed. There is no reasoning to decide when `this.value` reverts into `initialState` and when it reverts into `null`. Whatever we choose, developers will be stumbling on it and will be forced to study the addon's source code to see how `initialValue` actually affects `value`.

    With `initialValue` gone, developers are free to do `myResource.value || null` instead. This is no biggie and it makes the behavior 100% unambiguous to them and their teammates.

    Additionally, `ember-async-data` does not have it, and it's a good example to follow: no ambiguity means fewer surprises for developers.

***

As a result, the `trackedFunction` and `State` code is now delightfully small. `ember-async-data` takes care of a lot of things, now it's not neccessary to `waitForPromise`, to reassign promise state, there is no need for `notQuitePromise`. My hack of `await settled()` inside `retry` is also no longer necessary.

The removal of `initialState` also simplified the file drastically.

***

There is one failing test: `Utils | trackedFunction | rendering: it is retryable `.

This is a test where a component with a resource is rendered, then a `<button {{on 'click' this.data.retry}}>` is clicked. The expectation is that the resource's callback fires twice.

For some reason, when the button is clicked, `retry` is called correctly, but immediately afterwards the Resource Manager reruns the resource callback (as if a tangled tracked value has changed — but it hasn't). As a result, the `State` is recreated and `retry` is called for the third time.

I failed to figure out why it happens. Maybe you can help?

lolmaus commented on testing/ember-app/tests/utils/function/state-test.ts:
```diff
@@ -11,7 +11,7 @@ module('Utils | State | js', function (hooks) {
 
   test('initial state', async function (assert) {
-    let m;
+    let m = '<invalid state>';
```
Here `m` is used entirely to make test more readable by making the custom assertion message appear on its own line, rather than after the assertion where it's hard to notice and to distinguish from other arguments.

lolmaus commented on ember-resources/src/util/function.ts:
```diff
@@ -134,9 +137,29 @@ export class State<Value> {
+    this.promise = this.#fn();
+
+    // TrackedAsyncData interacts with tracked data during instantiation.
+    // We don't want this internal state to entangle with `trackedFunction`
```
😨 Now I get it.

NullVoxPopuli commented:
I broke the types, I'll fix in a bit

NullVoxPopuli commented:
Superseded by: https://github.com/NullVoxPopuli/ember-resources/pull/779
